### DISCUSSION TITLE
Is this intended for normie Linux user?

### DISCUSSION THREAD

**Discussion by slrslr** [hakavlad/le9-patch#16] [Q&A] [answered]
The readme does not answer question like:
How to measure current Linux system state without using this patch, what to run to apply this patch, how to measure the state after that - in order to understand the impact/difference before/after as a layman Linux GUI user. How to uninstall.
Maybe this is not intended for broad use by regular/normie Linux users who use basically GUI? Is there any other tool, which is more simple and does similar and allows measuring impact?

**Comment by hakavlad** [comment, marked as answer]
It's experimental and not for normies.

Just use MGLRU, it's merged in Linux 6.1.
https://www.phoronix.com/news/MGLRU-LPC-2022

**Reply by slrslr** [reply]
Thank You. I have read a bit about that and I am surprised it is not found on
wiki: https://wiki.debian.org/FrontPage?action=fullsearch&titlesearch=0&value=MGLRU&context=180
forum: https://forums.debian.net/search.php?keywords=MGLRU
apt: apt search MGLRU*
https://docs.kernel.org/search.html?q=MGLRU
My Debian 12 (stable) is running 6.1.0-27-amd64

But there are some articles:
https://dev.to/archerallstars/enable-the-magical-multi-gen-lru-aka-mglru-on-opensuse-tumblewed-stay-away-from-oom-today-ji6
https://docs.kernel.org/admin-guide/mm/multigen_lru.html
https://github.com/hakavlad/mg-lru-helper/blob/main/README.md

Which resulting in these possible steps:

Check whether MGLRU is supported by the kernel:

> grep -Ria CONFIG_LRU_GEN /boot; echo "Y means that it is supported. Otherwise do not continue."

A) enable in a grub configuration file:

command "sudo nano /etc/default/grub" and possibly insert "mglru.enabled=1" into a GRUB_CMDLINE_LINUX, save the file and run "sudo update-grub". Then the command "grep linuz /boot/grub/grub.cfg" should show this kernel parameter in a grub boot entries. If not or not in all, undo what you did and update grub again and then try method B bellow:

B) Install the service:

> sudo wget -q https://pastebin.com/raw/ADq1GETU -O /etc/systemd/system/mglru.service && sudo systemctl daemon-reload && sudo systemctl enable mglru && systemctl start mglru

Check the status:

> cat /sys/kernel/mm/lru_gen/enabled; echo "This should have returned 0x0007. If it returned 0x0000, this means it's not enabled yet."`

> cat /sys/kernel/mm/lru_gen/min_ttl_ms; echo "This should have returned 1000 for the optimal thrashing prevention."

> sudo journalctl -kg lru && cat /proc/mglru # some details about lru (after reboot) ?

read above mentioned links and find some ways to measure impact of it

**Reply by hakavlad** [reply]
Multigenerational LRU (MGLRU) is a memory management mechanism that enhances the performance of page caching in operating systems, particularly in scenarios with high memory pressure. It improves the efficiency of the Least Recently Used (LRU) algorithm by categorizing pages into multiple generations based on their usage patterns.

**In Debian 12 (Bookworm), MGLRU support is included in the kernel by default.** This feature helps in better management of memory by allowing the system to make more informed decisions about which pages to keep in memory and which to evict, ultimately leading to improved performance in memory-intensive applications.

To check if MGLRU is enabled in your current kernel, you can run the following command in the terminal:

bash

cat /boot/config-$(uname -r) | grep MGLRU

If you see a line containing CONFIG_MGLRU=y, it means that MGLRU support is enabled. If it says CONFIG_MGLRU is not set, then it is not enabled.

Overall, MGLRU is a valuable enhancement for memory management in modern Linux kernels, contributing to better performance and responsiveness in various workloads.

-- ChatGPT

**Reply by hakavlad** [reply]
See also https://github.com/hakavlad/mg-lru-helper
